File: controllers/controller.go
package controllers

import (
	"context"
	_ "embed"
	"encoding/json"
	"io"

	"kitties/models"
	"kitties/services"
)

//go:embed schemas/create_request.json
var CreateRequestSchema string

//go:embed schemas/update_request.json
var UpdateRequestSchema string

//go:embed schemas/replace_request.json
var ReplaceRequestSchema string

// Bounds for list pagination, protecting the datastore from unbounded reads.
const (
	DefaultListLimit = 100
	MaxListLimit     = 1000
)

// CoerceLimit clamps a requested page size into the supported range,
// falling back to the default when the value is missing or invalid.
func CoerceLimit(limit int) int {
	if limit < 1 {
		return DefaultListLimit
	}
	if limit > MaxListLimit {
		return MaxListLimit
	}
	return limit
}

type CatController interface {
	Get(ctx context.Context, id string) (*models.CatDto, error)
	GetList(ctx context.Context, limit int) ([]models.CatDto, error)
	Create(ctx context.Context, body io.Reader) (*models.CatDto, error)
	Update(ctx context.Context, id string, body io.Reader) (*models.CatDto, error)
	Delete(ctx context.Context, id string) error
}

type catController struct {
	service         services.CatService
	schemaValidator services.SchemaValidator
}

func NewCatController(service services.CatService, schemaValidator services.SchemaValidator) CatController {
	return &catController{service: service, schemaValidator: schemaValidator}
}

func (c *catController) Get(ctx context.Context, id string) (*models.CatDto, error) {
	return c.service.Get(ctx, id)
}

func (c *catController) GetList(ctx context.Context, limit int) ([]models.CatDto, error) {
	return c.service.GetList(ctx, CoerceLimit(limit))
}

func (c *catController) Create(ctx context.Context, body io.Reader) (*models.CatDto, error) {
	var req models.CreateCatRequest
	if err := json.NewDecoder(body).Decode(&req); err != nil {
		return nil, &models.ValidationError{Message: "Invalid request body."}
	}

	if err := c.schemaValidator.Validate(req, "create_request"); err != nil {
		return nil, err
	}

	return c.service.Create(ctx, req)
}

func (c *catController) Update(ctx context.Context, id string, body io.Reader) (*models.CatDto, error) {
	var req models.UpdateCatRequest
	if err := json.NewDecoder(body).Decode(&req); err != nil {
		return nil, &models.ValidationError{Message: "Invalid request body."}
	}

	req.Id = id

	if err := c.schemaValidator.Validate(req, "update_request"); err != nil {
		return nil, err
	}

	return c.service.Update(ctx, req)
}

func (c *catController) Delete(ctx context.Context, id string) error {
	return c.service.Delete(ctx, id)
}

type DogController interface {
	Get(ctx context.Context, id string) (*models.DogDto, error)
	GetList(ctx context.Context, limit int) ([]models.DogDto, error)
	Create(ctx context.Context, body io.Reader) (*models.DogDto, error)
	Replace(ctx context.Context, id string, body io.Reader) (*models.DogDto, error)
	Delete(ctx context.Context, id string) error
}

type dogController struct {
	service         services.DogService
	schemaValidator services.SchemaValidator
}

func NewDogController(service services.DogService, schemaValidator services.SchemaValidator) DogController {
	return &dogController{service: service, schemaValidator: schemaValidator}
}

func (c *dogController) Get(ctx context.Context, id string) (*models.DogDto, error) {
	return c.service.Get(ctx, id)
}

func (c *dogController) GetList(ctx context.Context, limit int) ([]models.DogDto, error) {
	return c.service.GetList(ctx, CoerceLimit(limit))
}

func (c *dogController) Create(ctx context.Context, body io.Reader) (*models.DogDto, error) {
	var req models.CreateDogRequest
	if err := json.NewDecoder(body).Decode(&req); err != nil {
		return nil, &models.ValidationError{Message: "Invalid request body."}
	}

	if err := c.schemaValidator.Validate(req, "create_request"); err != nil {
		return nil, err
	}

	return c.service.Create(ctx, req)
}

func (c *dogController) Replace(ctx context.Context, id string, body io.Reader) (*models.DogDto, error) {
	var req models.ReplaceDogRequest
	if err := json.NewDecoder(body).Decode(&req); err != nil {
		return nil, &models.ValidationError{Message: "Invalid request body."}
	}

	req.Id = id

	if err := c.schemaValidator.Validate(req, "replace_request"); err != nil {
		return nil, err
	}

	return c.service.Replace(ctx, req)
}

func (c *dogController) Delete(ctx context.Context, id string) error {
	return c.service.Delete(ctx, id)
}

File: controllers/controller.py
from typing import List, Optional

from flask import Request

from services import (
    CatService,
    DogService,
)
from repositories.repository import DEFAULT_LIST_LIMIT
from models import (
    CatResponse,
    Cat,
    CatIdValidation,
    DogResponse,
    Dog,
    DogIdValidation,
)

MAX_LIST_LIMIT = 1000


def _coerce_limit(raw: Optional[str]) -> int:
    try:
        limit = int(raw)
    except (TypeError, ValueError):
        return DEFAULT_LIST_LIMIT
    if limit < 1:
        return DEFAULT_LIST_LIMIT
    return min(limit, MAX_LIST_LIMIT)


class CatController:
    def __init__(self, service: CatService):
        self.service = service

    async def get_by_id(self, request: Request) -> CatResponse:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        CatIdValidation(id=item_id)
        return await self.service.get_by_id(item_id)

    async def get_list(self, request: Request) -> List[CatResponse]:
        limit = _coerce_limit(request.args.get('limit'))
        return await self.service.get_list(limit)

    async def create(self, request: Request) -> CatResponse:
        item_json: dict = request.get_json()
        item = Cat(**item_json)
        return await self.service.create(item)

    async def update(self, request: Request) -> CatResponse:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        item_data: dict = request.get_json()
        item = Cat(**item_data)
        item.id = item_id
        return await self.service.update(item)

    async def soft_delete(self, request: Request) -> None:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        await self.service.soft_delete(item_id)


class DogController:
    def __init__(self, service: DogService):
        self.service = service

    async def get_by_id(self, request: Request) -> DogResponse:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        DogIdValidation(id=item_id)
        return await self.service.get_by_id(item_id)

    async def get_list(self, request: Request) -> List[DogResponse]:
        limit = _coerce_limit(request.args.get('limit'))
        return await self.service.get_list(limit)

    async def create(self, request: Request) -> DogResponse:
        item_json: dict = request.get_json()
        item = Dog(**item_json)
        return await self.service.create(item)

    async def replace(self, request: Request) -> DogResponse:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        item_data: dict = request.get_json()
        item = Dog(**item_data)
        item.id = item_id
        return await self.service.replace(item)

    async def soft_delete(self, request: Request) -> None:
        path_parts = request.path.strip('/').split('/')
        item_id: str = path_parts[-1] if len(path_parts) > 0 else None
        await self.service.soft_delete(item_id)

